core/src/solve.jl

```diff
@@ -4,6 +4,57 @@ const ScalarInterpolation =
 const VectorInterpolation =
     LinearInterpolation{Vector{Vector{Float64}}, Vector{Float64}, true, Vector{Float64}}
 
+"""
+This struct is analogous to SparseArrays.SparseMatrixCSC, with the only
+difference that nzval is a DiffCache instead of a vector.
+"""
+struct SparseMatrixCSC_DiffCache{D <: DiffCache, Ti <: Integer}
+    m::Int                  # Number of rows
+    n::Int                  # Number of columns
+    colptr::Vector{Ti}      # Column j is in colptr[j]:(colptr[j+1]-1)
+    rowval::Vector{Ti}      # Row indices of stored values
+    nzval::D                # DiffCache
+end
+
+"""
+This struct is analogous to SparseArrays.SparseMatrixCSC, with the only
+difference that nzval is a ReinterpretArray (of Dual numbers).
+
+SparseMatrixCSC_DiffCache and SparseMatrixCSC_cache are used to have a multi-level DiffCache of a sparse matrix
+whose cache (accessed via get_tmp_sparse) supports sparse indexing.
+Previously FixedSizeDiffCache was used for the sparse matrix Connectivity.flow, but FixedSizeDiffCache
+does not support multi-level Duals.
+"""
+struct SparseMatrixCSC_cache{T <: Number, Ti <: Integer} <:
+       SparseArrays.AbstractSparseMatrixCSC{T, Ti}
+    m::Int                  # Number of rows
+    n::Int                  # Number of columns
+    colptr::Vector{Ti}      # Column j is in colptr[j]:(colptr[j+1]-1)
+    rowval::Vector{Ti}      # Row indices of stored values
+    nzval::Base.ReinterpretArray{T} # cache
```

Comment:
This field is not yet concretely typed, since a ReinterpretArray has many more type parameters: https://github.com/JuliaLang/julia/blob/v1.10.0-beta3/base/reinterpretarray.jl#L9

In combination with https://github.com/Deltares/Ribasim/issues/694#issuecomment-1775409127 I'm starting to wonder if we should keep using a sparse matrix for flow at all, or if we should use some other data structure like a dict.